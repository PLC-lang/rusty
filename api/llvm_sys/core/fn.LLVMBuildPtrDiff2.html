<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `LLVMBuildPtrDiff2` fn in crate `llvm_sys`."><title>LLVMBuildPtrDiff2 in llvm_sys::core - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-dd39b87e5fcfba68.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="llvm_sys" data-themes="" data-resource-suffix="" data-rustdoc-version="1.80.1 (3f5fd8dd4 2024-08-06)" data-channel="1.80.1" data-search-js="search-d52510db62a78183.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-118b08c4c78b968e.js"></script><script defer src="sidebar-items.js"></script><script defer src="../../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-df360f571f6edeae.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc fn"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../llvm_sys/index.html">llvm_sys</a><span class="version">140.1.3</span></h2></div><div class="sidebar-elems"><h2><a href="index.html">In llvm_sys::core</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Function <a href="../index.html">llvm_sys</a>::<wbr><a href="index.html">core</a>::<wbr><a class="fn" href="#">LLVMBuildPtrDiff2</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../src/llvm_sys/core.rs.html#2002-2008">source</a> Â· <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>pub unsafe extern &quot;C&quot; fn LLVMBuildPtrDiff2(
    arg1: <a class="type" href="../prelude/type.LLVMBuilderRef.html" title="type llvm_sys::prelude::LLVMBuilderRef">LLVMBuilderRef</a>,
    ElemTy: <a class="type" href="../prelude/type.LLVMTypeRef.html" title="type llvm_sys::prelude::LLVMTypeRef">LLVMTypeRef</a>,
    LHS: <a class="type" href="../prelude/type.LLVMValueRef.html" title="type llvm_sys::prelude::LLVMValueRef">LLVMValueRef</a>,
    RHS: <a class="type" href="../prelude/type.LLVMValueRef.html" title="type llvm_sys::prelude::LLVMValueRef">LLVMValueRef</a>,
    Name: <a class="primitive" href="https://doc.rust-lang.org/1.80.1/std/primitive.pointer.html">*const </a><a class="type" href="../../libc/unix/linux_like/linux/gnu/b64/x86_64/type.c_char.html" title="type libc::unix::linux_like::linux::gnu::b64::x86_64::c_char">c_char</a>,
) -&gt; <a class="type" href="../prelude/type.LLVMValueRef.html" title="type llvm_sys::prelude::LLVMValueRef">LLVMValueRef</a></code></pre></section></div></main></body></html>