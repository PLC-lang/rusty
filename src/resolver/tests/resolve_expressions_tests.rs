use core::panic;

use crate::{
    ast::{DataType, Statement, UserTypeDeclaration},
    index::Index,
    resolver::{
        tests::{annotate, parse},
        AnnotationMap, StatementAnnotation,
    },
    typesystem::VOID_TYPE,
};

#[test]
fn binary_expressions_resolves_types() {
    let (unit, index) = parse(
        "PROGRAM PRG
            1 + 2;
            1 + 2000;
            2147483648 + 1;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec!["DINT", "DINT", "LINT"];

    let types: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(expected_types, types);
}

#[test]
fn unary_expressions_resolves_types() {
    let (unit, index) = parse(
        "PROGRAM PRG
            NOT TRUE;
            -(2+3);
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec!["BOOL", "DINT"];

    let types: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(expected_types, types);
}

#[test]
fn binary_expressions_resolves_types_with_floats() {
    let (unit, index) = parse(
        "PROGRAM PRG
            1 + 2.2;
            1.1 + 2000;
            2000.0 + 1.0;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec!["REAL", "REAL", "REAL"];
    for (i, s) in statements.iter().enumerate() {
        assert_eq!(
            expected_types[i],
            annotations.get_type_or_void(s, &index).get_name(),
            "{:#?}",
            s
        );
    }
}

#[test]
fn local_variables_resolves_types() {
    let (unit, index) = parse(
        "PROGRAM PRG
            VAR
                b : BYTE;
                w : WORD;
                dw : DWORD;
                lw : LWORD;
                si : SINT;
                usi : USINT;
                i : INT;
                ui : UINT;
                di : DINT;
                udi : UDINT;
                li : LINT;
                uli : ULINT;
            END_VAR

            b;
            w;
            dw;
            lw;
            si;
            usi;
            i;
            ui;
            di;
            udi;
            li;
            uli;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec![
        "BYTE", "WORD", "DWORD", "LWORD", "SINT", "USINT", "INT", "UINT", "DINT", "UDINT", "LINT",
        "ULINT",
    ];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn global_resolves_types() {
    let (unit, index) = parse(
        "
        VAR_GLOBAL
            b : BYTE;
            w : WORD;
            dw : DWORD;
            lw : LWORD;
            si : SINT;
            usi : USINT;
            i : INT;
            ui : UINT;
            di : DINT;
            udi : UDINT;
            li : LINT;
            uli : ULINT;
        END_VAR
        
        PROGRAM PRG
            b;
            w;
            dw;
            lw;
            si;
            usi;
            i;
            ui;
            di;
            udi;
            li;
            uli;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec![
        "BYTE", "WORD", "DWORD", "LWORD", "SINT", "USINT", "INT", "UINT", "DINT", "UDINT", "LINT",
        "ULINT",
    ];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn resolve_binary_expressions() {
    let (unit, index) = parse(
        "
        VAR_GLOBAL
            b : BYTE;
            w : WORD;
            dw : DWORD;
            lw : LWORD;
            si : SINT;
            usi : USINT;
            i : INT;
            ui : UINT;
            di : DINT;
            udi : UDINT;
            li : LINT;
            uli : ULINT;
        END_VAR
        
        PROGRAM PRG
            b + b;
            b + w;
            b + dw;
            b + lw;
            b + si;
            b + usi;
            b + i;
            b + ui;
            b + di;
            b + udi;
            b + li;
            b + uli;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec![
        "BYTE", "WORD", "DWORD", "LWORD", "SINT", "BYTE", "INT", "UINT", "DINT", "UDINT", "LINT",
        "ULINT",
    ];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn complex_expressions_resolve_types() {
    let (unit, index) = parse(
        "PROGRAM PRG
            VAR
                b : BYTE;
                w : WORD;
                dw : DWORD;
                lw : LWORD;
                si : SINT;
                usi : USINT;
                i : INT;
                ui : UINT;
                di : DINT;
                udi : UDINT;
                li : LINT;
                uli : ULINT;
                r : REAL;
            END_VAR

            b + w * di + li;
            b + w + di;
            b + w * di + r;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec!["LINT", "DINT", "REAL"];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn array_expressions_resolve_types() {
    let (unit, index) = parse(
        "PROGRAM PRG
            VAR
                i : ARRAY[0..10] OF INT;
                y : ARRAY[0..10] OF MyInt;
                a : MyIntArray;
                b : MyAliasArray;
            END_VAR

            i;
            i[2];

            y;
            y[2];

            a;
            a[2];

            b;
            b[2];
        END_PROGRAM
        
        TYPE MyInt: INT := 7; END_TYPE 
        TYPE MyIntArray: ARRAY[0..10] OF INT := 7; END_TYPE 
        TYPE MyAliasArray: ARRAY[0..10] OF MyInt := 7; END_TYPE 

        ",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec![
        "__PRG_i",
        "INT",
        "__PRG_y",
        "INT",
        "MyIntArray",
        "INT",
        "MyAliasArray",
        "INT",
    ];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn qualified_expressions_resolve_types() {
    let (unit, index) = parse(
        "
         PROGRAM Other
            VAR_INPUT
                b : BYTE;
                w : WORD;
                dw : DWORD;
                lw : LWORD;
            END_VAR
        END_PROGRAM   

        PROGRAM PRG
            Other.b;
            Other.w;
            Other.dw;
            Other.lw;
            Other.b + Other.w;
            Other.b + Other.w + Other.dw;
            Other.b + Other.w + Other.dw + Other.lw;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[1].statements;

    let expected_types = vec!["BYTE", "WORD", "DWORD", "LWORD", "WORD", "DWORD", "LWORD"];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn pou_expressions_resolve_types() {
    let (unit, index) = parse(
        "
        PROGRAM OtherPrg
        END_PROGRAM   

        FUNCTION OtherFunc : INT
        END_FUNCTION

        FUNCTION_BLOCK OtherFuncBlock
        END_FUNCTION_BLOCK

        PROGRAM PRG
            OtherPrg;
            OtherFunc;
            OtherFuncBlock;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[3].statements;

    //none of these pou's should really resolve to a type
    let expected_types = vec![VOID_TYPE, VOID_TYPE, VOID_TYPE];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();
    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));

    assert_eq!(
        Some(&StatementAnnotation::ProgramAnnotation {
            qualified_name: "OtherPrg".into()
        }),
        annotations.get_annotation(&statements[0])
    );
    assert_eq!(
        Some(&StatementAnnotation::FunctionAnnotation {
            qualified_name: "OtherFunc".into(),
            return_type: "INT".into()
        }),
        annotations.get_annotation(&statements[1])
    );
    assert_eq!(
        Some(&StatementAnnotation::TypeAnnotation {
            type_name: "OtherFuncBlock".into()
        }),
        annotations.get_annotation(&statements[2])
    );
}

#[test]
fn assignment_expressions_resolve_types() {
    let (unit, index) = parse(
        "
        PROGRAM PRG
            VAR
                x : INT;
                y : BYTE;
                z : LWORD;
            END_VAR

            x := y;
            z := x;
        END_PROGRAM",
    );
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec![VOID_TYPE, VOID_TYPE];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));

    if let Statement::Assignment { left, right, .. } = &statements[0] {
        assert_eq!(
            annotations.get_type_or_void(&left, &index).get_name(),
            "INT"
        );
        assert_eq!(
            annotations.get_type_or_void(&right, &index).get_name(),
            "BYTE"
        );
    } else {
        panic!("expected assignment")
    }
    if let Statement::Assignment { left, right, .. } = &statements[1] {
        assert_eq!(
            annotations.get_type_or_void(&left, &index).get_name(),
            "LWORD"
        );
        assert_eq!(
            annotations.get_type_or_void(&right, &index).get_name(),
            "INT"
        );
    } else {
        panic!("expected assignment")
    }
}

#[test]
fn qualified_expressions_to_structs_resolve_types() {
    let (unit, index) = parse(
        "
        TYPE NextStruct: STRUCT
            b : BYTE;
            w : WORD;
            dw : DWORD;
            lw : LWORD;
        END_STRUCT
        END_TYPE
 
        TYPE MyStruct: STRUCT
            b : BYTE;
            w : WORD;
            dw : DWORD;
            lw : LWORD;
            next : NextStruct;
        END_STRUCT
        END_TYPE

        PROGRAM PRG
            VAR 
                mys : MyStruct;
            END_VAR
            mys;
            mys.b;
            mys.w;
            mys.dw;
            mys.lw;
            mys.next;
            mys.next.b;
            mys.next.w;
            mys.next.dw;
            mys.next.lw;
        END_PROGRAM",
    );

    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec![
        "MyStruct",
        "BYTE",
        "WORD",
        "DWORD",
        "LWORD",
        "NextStruct",
        "BYTE",
        "WORD",
        "DWORD",
        "LWORD",
    ];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn qualified_expressions_to_inlined_structs_resolve_types() {
    let (unit, index) = parse(
        "
        PROGRAM PRG
            VAR 
                mys : STRUCT
                    b : BYTE;
                    w : WORD;
                    dw : DWORD;
                    lw : LWORD;
                END_STRUCT;
            END_VAR
            mys;
            mys.b;
            mys.w;
            mys.dw;
            mys.lw;
        END_PROGRAM",
    );

    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec!["__PRG_mys", "BYTE", "WORD", "DWORD", "LWORD"];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn function_expression_resolves_to_the_function_itself_not_its_return_type() {
    //GIVEN a reference to a function
    let (unit, index) = parse(
        "
        FUNCTION foo : INT
        END_FUNCTION

        PROGRAM PRG
            foo;
        END_PROGRAM
        ",
    );

    //WHEN the AST is annotated
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[1].statements;

    // THEN we expect it to be annotated with the function itself
    let foo_annotation = annotations.get_annotation(&statements[0]);
    assert_eq!(
        Some(&StatementAnnotation::FunctionAnnotation {
            qualified_name: "foo".into(),
            return_type: "INT".into()
        }),
        foo_annotation
    );

    // AND we expect no type to be associated with the expression
    let associated_type = annotations.get_type(&statements[0], &index);
    assert_eq!(None, associated_type);
}

#[test]
fn function_call_expression_resolves_to_the_function_itself_not_its_return_type() {
    //GIVEN a reference to a function
    let (unit, index) = parse(
        "
        FUNCTION foo : INT
        END_FUNCTION

        PROGRAM PRG
            foo();
        END_PROGRAM
        ",
    );

    //WHEN the AST is annotated
    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[1].statements;

    // THEN we expect it to be annotated with the function itself
    let foo_annotation = annotations.get_annotation(&statements[0]);
    assert_eq!(
        Some(&StatementAnnotation::ExpressionAnnotation {
            resulting_type: "INT".into()
        }),
        foo_annotation
    );

    // AND we expect no type to be associated with the expression
    let associated_type = annotations.get_type(&statements[0], &index);
    assert_eq!(index.find_type("INT"), associated_type);
}

#[test]
fn qualified_expressions_to_aliased_structs_resolve_types() {
    let (unit, index) = parse(
        "
        TYPE NextStruct: STRUCT
            b : BYTE;
            w : WORD;
            dw : DWORD;
            lw : LWORD;
        END_STRUCT
        END_TYPE
 
        TYPE MyStruct: STRUCT
            b : BYTE;
            w : WORD;
            dw : DWORD;
            lw : LWORD;
            next : AliasedNextStruct;
        END_STRUCT
        END_TYPE

        TYPE AliasedMyStruct : MyStruct; END_TYPE
        TYPE AliasedNextStruct : NextStruct; END_TYPE

        PROGRAM PRG
            VAR 
                mys : AliasedMyStruct;
            END_VAR
            mys;
            mys.b;
            mys.w;
            mys.dw;
            mys.lw;
            mys.next;
            mys.next.b;
            mys.next.w;
            mys.next.dw;
            mys.next.lw;
        END_PROGRAM",
    );

    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    let expected_types = vec![
        "MyStruct",
        "BYTE",
        "WORD",
        "DWORD",
        "LWORD",
        "NextStruct",
        "BYTE",
        "WORD",
        "DWORD",
        "LWORD",
    ];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn qualified_expressions_to_fbs_resolve_types() {
    let (unit, index) = parse(
        "
        FUNCTION_BLOCK MyFb
            VAR_INPUT
                fb_b : SINT;
                fb_i : INT;
                fb_d : DINT;
            END_VAR
        END_FUNCTION_BLOCK

        PROGRAM PRG
            VAR 
                fb : MyFb;
            END_VAR
            fb;
            fb.fb_b;
            fb.fb_i;
            fb.fb_d;
       END_PROGRAM",
    );

    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[1].statements;

    let expected_types = vec!["MyFb", "SINT", "INT", "DINT"];
    let type_names: Vec<&str> = statements
        .iter()
        .map(|s| annotations.get_type_or_void(s, &index).get_name())
        .collect();

    assert_eq!(format!("{:?}", expected_types), format!("{:?}", type_names));
}

#[test]
fn qualified_expressions_dont_fallback_to_globals() {
    let (unit, index) = parse(
        "
        VAR_GLOBAL
            x : DINT;
        END_VAR 

        TYPE MyStruct: STRUCT
            y : INT;
        END_STRUCT
        END_TYPE

        PROGRAM PRG
            VAR P : MyStruct; END_VAR
            P.x;
            P.y;
        END_PROGRAM",
    );

    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[0].statements;

    assert_eq!(None, annotations.get_annotation(&statements[0]));
    assert_eq!(
        Some(&StatementAnnotation::VariableAnnotation {
            qualified_name: "MyStruct.y".into(),
            resulting_type: "INT".into()
        }),
        annotations.get_annotation(&statements[1])
    );
}

#[test]
fn function_parameter_assignments_resolve_types() {
    let (unit, index) = parse(
        "
        FUNCTION foo : MyType
            VAR_INPUT
                x : INT;
            END_VAR
            VAR_OUTPUT
                y : SINT;
            END_VAR
        END_FUNCTION

        PROGRAM PRG
            foo(x := 3, y => 6);
        END_PROGRAM
        
        TYPE MyType: INT; END_TYPE
        ",
    );

    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[1].statements;

    assert_eq!(
        annotations
            .get_type_or_void(&statements[0], &index)
            .get_name(),
        "INT"
    );
    assert_eq!(
        annotations.get_annotation(&statements[0]),
        Some(&StatementAnnotation::expression("INT"))
    );
    if let Statement::CallStatement {
        operator,
        parameters,
        ..
    } = &statements[0]
    {
        //make sure the call's operator resolved correctly
        assert_eq!(
            annotations.get_type_or_void(operator, &index).get_name(),
            VOID_TYPE
        );
        assert_eq!(
            annotations.get_annotation(operator),
            Some(&StatementAnnotation::FunctionAnnotation {
                qualified_name: "foo".into(),
                return_type: "MyType".into()
            })
        );

        if let Some(Statement::ExpressionList { expressions, .. }) = &**parameters {
            if let Statement::Assignment { left, right, .. } = &expressions[0] {
                assert_eq!(annotations.get_type_or_void(left, &index).get_name(), "INT");
                assert_eq!(
                    annotations.get_type_or_void(right, &index).get_name(),
                    "DINT"
                );
            } else {
                panic!("assignment expected")
            }
            if let Statement::OutputAssignment { left, right, .. } = &expressions[1] {
                assert_eq!(
                    annotations.get_type_or_void(left, &index).get_name(),
                    "SINT"
                );
                assert_eq!(
                    annotations.get_type_or_void(right, &index).get_name(),
                    "DINT"
                );
            } else {
                panic!("assignment expected")
            }
        } else {
            panic!("expression list expected")
        }
    } else {
        panic!("call statement");
    }
}

#[test]
fn nested_function_parameter_assignments_resolve_types() {
    let (unit, index) = parse(
        "
        FUNCTION foo : INT
            VAR_INPUT
                x : INT;
                y : BOOL;
            END_VAR
        END_FUNCTION

        FUNCTION baz : DINT
            VAR_INPUT
                x : DINT;
                y : DINT;
            END_VAR
        END_FUNCTION


        PROGRAM PRG
            VAR r: REAL; END_VAR
            foo(x := baz(x := 200, y := FALSE), y := baz(x := 200, y := TRUE) + r);
        END_PROGRAM",
    );

    let annotations = annotate(&unit, &index);
    let statements = &unit.implementations[2].statements;
    if let Statement::CallStatement { parameters, .. } = &statements[0] {
        //check the two parameters
        assert_parameter_assignment(parameters, 0, "INT", "DINT", &annotations, &index);
        assert_parameter_assignment(parameters, 1, "BOOL", "REAL", &annotations, &index);

        //check the inner call in the first parameter assignment of the outer call `x := baz(...)`
        if let Statement::Assignment { right, .. } = get_expression_from_list(parameters, 0) {
            if let Statement::CallStatement { parameters, .. } = right.as_ref() {
                // the left side here should be `x` - so lets see if it got mixed up with the outer call's `x`
                assert_parameter_assignment(parameters, 0, "DINT", "DINT", &annotations, &index);
            } else {
                panic!("inner call")
            }
        } else {
            panic!("assignment");
        }
    } else {
        panic!("call statement")
    }
}

#[test]
fn type_initial_values_are_resolved() {
    let (unit, index) = parse(
        "
        TYPE MyStruct : STRUCT
            x : INT := 20;
            y : BOOL := TRUE;
            z : STRING := 'abc';
        END_STRUCT
        END_TYPE
        ",
    );

    let annotations = annotate(&unit, &index);
    let UserTypeDeclaration { data_type, .. } = &unit.types[0];

    if let DataType::StructType { variables, .. } = data_type {
        assert_eq!(
            Some(&StatementAnnotation::expression("DINT")),
            annotations.get(variables[0].initializer.as_ref().unwrap())
        );
        assert_eq!(
            Some(&StatementAnnotation::expression("BOOL")),
            annotations.get(variables[1].initializer.as_ref().unwrap())
        );
        assert_eq!(
            Some(&StatementAnnotation::expression("STRING")),
            annotations.get(variables[2].initializer.as_ref().unwrap())
        );
    } else {
        panic!("no datatype: {:#?}", data_type)
    }
}

fn get_expression_from_list(stmt: &Option<Statement>, index: usize) -> &Statement {
    if let Some(Statement::ExpressionList { expressions, .. }) = stmt {
        &expressions[index]
    } else {
        panic!("no expression_list, found {:#?}", stmt)
    }
}

fn assert_parameter_assignment(
    parameters: &Option<Statement>,
    param_index: usize,
    left_type: &str,
    right_type: &str,
    annotations: &AnnotationMap,
    index: &Index,
) {
    if let Some(Statement::ExpressionList { expressions, .. }) = parameters {
        if let Statement::Assignment { left, right, .. } = &expressions[param_index] {
            assert_eq!(
                annotations.get_type_or_void(left, index).get_name(),
                left_type
            );
            assert_eq!(
                annotations.get_type_or_void(right, index).get_name(),
                right_type
            );
        } else {
            panic!("assignment expected")
        }
    } else {
        panic!("expression list expected")
    }
}
