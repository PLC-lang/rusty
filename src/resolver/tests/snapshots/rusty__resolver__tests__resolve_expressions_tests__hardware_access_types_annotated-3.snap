---
source: src/resolver/tests/resolve_expressions_tests.rs
expression: annotations.get(right).unwrap()
---
Variable {
    resulting_type: "DINT",
    qualified_name: "__M_1_3",
    constant: false,
    argument_type: ByVal(
        Global,
    ),
    auto_deref: None,
}
